### PR TITLE
Adding weighted loss functionality to L2Loss and CostVData

### PR DESCRIPTION
**Pull request by Ayush-iitkgp**
Adding weighted loss functionality to L2Loss and CostVData

### CODE REVIEW

File: src/cost_functions.jl [hunk 1]
```diff
@@ -36,16 +39,20 @@ struct L2Loss{T,D} <: DECostFunction
...
     push!(sol.u,fill(Inf,size(sol[1])))
   end
+  if weight == nothing
+    weight = ones(f.data)
+  end
   sumsq = 0.0
   @inbounds for i in 1:length(sol)
     for j in 1:length(sol[i])
       sumsq += (data[j,i] - sol[j,i])^2
+      sumsq *=weight[j,i]
```
Comment (ChrisRackauckas) [inline src/cost_functions.jl]: this will apply the weight to all previous values

**Comment by ChrisRackauckas** [comment]
I think weights should just be added as part of those loss functions, not to the objective function itself. The thing which shows that is the fact that this code directly sends it to the loss function and no where else.

**Comment by Ayush-iitkgp** [comment]
The default value of weight would be the matrix of ones which have the same size as of the data. Hence, it can't be an optional argument to the constructor. Would it be a right idea to ask the default values to the user?

**Comment by ChrisRackauckas** [comment]
Why not just default `nothing` and then handle that? I would make the no-weight case have less operations anyways.

**Comment by ChrisRackauckas** [comment]
Wait, why can't it be a default arg? One of the previous args has to be the data, so you can use that.

**Comment by ChrisRackauckas** [comment]
Needs a test.

**Comment by Ayush-iitkgp** [comment]
Test added and passing on Mac.

File: test/tests_on_odes/weighted_loss_test.jl [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+  tmp = rand()
+  weight[i,j] = tmp*tmp
+  d = Normal(0,tmp)
+  error[i,j] = rand(d)
+ end
+end
+
+weighted_data = original_solution_matrix_form + error
+
+weighted_cost_function = build_loss_objective(prob1,Tsit5(),CostVData(t,weighted_data',weight=weight'),maxiters=10000)
```
Comment (ChrisRackauckas) [inline test/tests_on_odes/weighted_loss_test.jl]: why the transposes?

**Pull request merged**